nautobot/core/tests/test_graphql.py

```diff
@@ -1620,43 +1674,81 @@ def test_query_interfaces_filter_second_level(self):
         """Test "second-level" filtering of Interfaces within a Devices query."""
 
         filters = (
-            (f'device_id: "{self.device1.id}"', 2),
-            ('kind: "virtual"', 2),
-            ('mac_address: "00:11:11:11:11:11"', 1),
-            ("vlan: 100", 1),
-            (f'vlan_id: "{self.vlan1.id}"', 1),
+            (
+                f'device_id: "{self.device1.id}"',
+                Q(device=self.device1),
+            ),
+            (
+                'kind: "virtual"',
+                Q(type=InterfaceTypeChoices.TYPE_VIRTUAL),
+            ),
+            (
+                'mac_address: "00:11:11:11:11:11"',
+                Q(mac_address="00:11:11:11:11:11"),
+            ),
+            (
+                "vlan: 100",
+                Q(untagged_vlan__vid=100) | Q(tagged_vlans__vid=100),
+            ),
+            (
+                f'vlan_id: "{self.vlan1.id}"',
+                Q(untagged_vlan=self.vlan1) | Q(tagged_vlans=self.vlan1),
+            ),
         )
 
-        for filterv, nbr_expected_results in filters:
-            with self.subTest(msg=f"Checking {filterv}", filterv=filterv, nbr_expected_results=nbr_expected_results):
-                query = "query { devices{ interfaces(" + filterv + "){ id }}}"
+        for filterv, qs_filter in filters:
+            with self.subTest(msg=f"Checking {filterv}", filterv=filterv, qs_filter=qs_filter):
+                matched = 0
+                query = "query { devices{ id, interfaces(" + filterv + "){ id }}}"
                 result = self.execute_query(query)
                 self.assertIsNone(result.errors)
-                self.assertEqual(len(result.data["devices"][0]["interfaces"]), nbr_expected_results)
+                for device in result.data["devices"]:
+                    qs = Interface.objects.filter(device_id=device["id"])
+                    expected_count = qs.filter(qs_filter).count()
+                    matched = max(matched, len(device["interfaces"]))
+                    self.assertEqual(len(device["interfaces"]), expected_count)
+                self.assertNotEqual(matched, 0, msg="At least one object matched GraphQL query")
```

Comment:
Might consider moving all of this repeated logic into a helper function, something that could be called like:

```python
self.validate_graphql_filters(filters, query_start="query { devices{ id, interfaces(", query_end="){ id }}}")
```

?

Reply:
We would have to pass in the `query` from line 1702, the `qs` from line 1706 and the key names in lines 1708-1709. I tried combining the logic and it made the tests more confusing to read and only saved a few lines of code. I recommend leaving them as is for now.